cmd/promtool/backfill_test.go

```diff
@@ -0,0 +1,441 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"io/ioutil"
+	"math"
+	"os"
+	"sort"
+	"testing"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/stretchr/testify/require"
+)
+
+type backfillSample struct {
+	Timestamp int64
+	Value     float64
+	Labels    labels.Labels
+}
+
+func createTemporaryOpenMetricsFile(t *testing.T, text string) string {
+	newf, err := ioutil.TempFile("", "")
+	require.NoError(t, err)
+
+	_, err = newf.WriteString(text)
+	require.NoError(t, err)
+	require.NoError(t, newf.Close())
+
+	return newf.Name()
+}
+
+func sortSamples(samples []backfillSample) {
+	sort.Slice(samples, func(x, y int) bool {
+		sx, sy := samples[x], samples[y]
+		if sx.Timestamp != sy.Timestamp {
+			return sx.Timestamp < sy.Timestamp
+		}
+		return sx.Value < sy.Value
+	})
+}
+
+func queryblock(t testing.TB, q storage.Querier, expectedMinTime, expectedMaxTime int64, lbls labels.Labels, matchers ...*labels.Matcher) ([]backfillSample, error) {
+	ss := q.Select(false, nil, matchers...)
+	defer func() {
+		require.NoError(t, q.Close())
+	}()
+	samples := []backfillSample{}
+	for ss.Next() {
+		series := ss.At()
+		it := series.Iterator()
+		require.NoError(t, it.Err())
+		for it.Next() {
+			ts, v := it.At()
+			samples = append(samples, backfillSample{Timestamp: ts, Value: v, Labels: series.Labels()})
+		}
+		if len(samples) == 0 {
+			continue
+		}
```

Comment:
yes, below we are printing blocks and I am printing the last block (len(blocks)-1) every time so that I can show progress. if len ==0, index will be -1 which will cause problem

Reply:
We don't :P